code/model3deeper.py

```diff
@@ -0,0 +1,161 @@
+import numpy as np
+import time
+import torch
+import torch.autograd as autograd
+import torch.nn as nn
+import torch.optim as optim
+
+import constants
+import plots
+import utils
+import dal
+import metrics
+
+from model import Model
+
+
+class Model3Deeper(nn.Module, Model):
+    """
+    Contains the code for Model3.
+    Architecture - sr_embedding -> linear_layer -> relu_activation -> linear_layer -> predicted_email_emb -> loss
+     - One sr_embedding is calculated for each email as a concatenation of sender_emb and average of all receiver_
+     emb present in the mail
+     - Loss is calculated as L2 loss between predicted_email_embedding and email_representation. The email
+     representation is obtained by averaging embeddings from pre-trained word2vec model
+    """
+
+    def __init__(self, pre_trained=False, load_from=None, hidden_dims=[500], use_batchnorm=False):
+        self.val_after_epoch = 1
+        self.best_val = 0.0
+        # keeps track of how many times the model has seen each email_id, either as a sender or receiver
+        self.emailid_train_freq = {}
+        super(Model3Deeper, self).__init__()
+        # embedding lookup for 150 users each have constants.USER_EMB_SIZE dimension representation
+        self.embedding_layer = nn.Embedding(150, constants.USER_EMB_SIZE)
+        # first hidden layer, linear layer with weights 2*constants.USER_EMB_SIZEx500
+        # this should be the size of <sender+receiver representation>
+        inp_dim = 2 * constants.USER_EMB_SIZE
+        self.hidden_layers, out_dim = self.create_hidden_layers(inp_dim, hidden_dims, use_batchnorm)
+        # final linear layer that outputs the predicted email representation
+        self.email_layer = nn.Linear(out_dim, constants.EMAIL_EMB_SIZE)
+        if pre_trained:
+            self.load(load_from)
+
+    def create_hidden_layers(self, inp_dim, hidden_dims, use_batchnorm):
```

Comment:
Can we pull this out as a common helper function?

Reply:
yep it can be moved to models.py ... i'll move it